stdlib/2and3/ctypes/__init__.pyi

```diff
@@ -0,0 +1,255 @@
+# Stubs for ctypes
+
+from typing import (
+    Any, Callable, ClassVar, Iterable, List, Mapping, Optional, Sequence, Sized, Text, Tuple, Type,
+    Generic, TypeVar, overload,
+)
+from typing import Union as _UnionT
+import sys
+
+_T = TypeVar('_T')
+_DLLT = TypeVar('_DLLT', bound=CDLL)
+_CT = TypeVar('_CT', bound=_CData)
+
+
+RTLD_GLOBAL: int = ...
+RTLD_LOCAL: int = ...
+DEFAULT_MODE: int = ...
+
+
+class CDLL(object):
+    _func_flags_: ClassVar[int] = ...
+    _func_restype_: ClassVar[_CData] = ...
+    _name: str = ...
+    _handle: int = ...
+    _FuncPtr: Type[_FuncPointer] = ...
+    def __init__(self, name: str, mode: int = ..., handle: Optional[int] = ...,
+                 use_errno: bool = ..., use_last_error: bool = ...) -> None: ...
+    def __getattr__(self, name: str) -> _FuncPointer: ...
+    def __getitem__(self, name: str) -> _FuncPointer: ...
+if sys.platform == 'win32':
+    class OleDLL(CDLL): ...
+    class WinDLL(CDLL): ...
+class PyDLL(CDLL): ...
+
+class LibraryLoader(Generic[_DLLT]):
+    def __init__(self, dlltype: Type[_DLLT]) -> None: ...
+    def __getattr__(self, name: str) -> _DLLT: ...
+    def __getitem__(self, name: str) -> _DLLT: ...
+    def LoadLibrary(self, name: str) -> _DLLT: ...
+
+cdll: LibraryLoader[CDLL] = ...
+if sys.platform == 'win32':
+    windll: LibraryLoader[WinDLL] = ...
+    oledll: LibraryLoader[OleDLL] = ...
+pydll: LibraryLoader[PyDLL] = ...
+pythonapi: PyDLL = ...
+
+class _CDataMeta(type):
+    # TODO The return type is not accurate. The method definition *should* look like this:
+    # def __mul__(cls: Type[_CT], other: int) -> Type[Array[_CT]]: ...
+    # but that is not valid, because technically a _CDataMeta might not be a Type[_CT].
+    # This can never actually happen, because all _CDataMeta instances are _CData subclasses, but a typechecker doesn't know that.
+    def __mul__(cls: _CDataMeta, other: int) -> Type[Array[_CT]]: ...
+    def __rmul__(cls: _CDataMeta, other: int) -> Type[Array[_CT]]: ...
+class _CData(metaclass=_CDataMeta):
+    _b_base: int = ...
+    _b_needsfree_: bool = ...
+    _objects: Optional[Mapping[Any, int]] = ...
+    @classmethod
+    def from_buffer(cls: Type[_CT], source: bytearray, offset: int = ...) -> _CT: ...
+    @classmethod
+    def from_buffer_copy(cls: Type[_CT], source: bytearray, offset: int = ...) -> _CT: ...
```

Comment:
Shouldn't this be `bytes` not `bytearray`?

Reply:
Both of the `bytearray` types are inadequate actually - `from_buffer_copy` accepts any object that implements the [buffer protocol](https://docs.python.org/3/c-api/buffer.html), and similarly `from_buffer` accepts any writable buffer. `bytearray` and `bytes` are common examples of read-write and read-only buffer objects, but there are others like `memoryview` and `array.array`. All `ctypes` data objects (`_CData` subclasses) are writable buffers as well.

Does typeshed/`typing` have a type for "anything implementing the buffer protocol"? If not, we probably need to use `Union`s and list the most common stdlib buffer types.

Reply:
You can define a new protocol. I think there are already many examples in typeshed (grep for `Protocol`).

Reply:
Oh, it is C-level, in this case looks like a big union is the only option.

Reply:
Isn't `bytes` already treated by mypy as anything implementing the buffer protocol?

Reply:
Fix here: #2610